presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java

```diff
@@ -0,0 +1,193 @@
+    {
+        this.taskManager = requireNonNull(taskManager, "taskManager is null");
+        this.responseExecutor = requireNonNull(responseExecutor, "responseExecutor is null");
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor is null");
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws IOException
+    {
+        String requestURI = request.getRequestURI();
+        // Example:  /v1/task/async/{taskId}/results/{bufferId}/{token}
+        List<String> requestURIParts = Arrays.asList(requestURI.split("/"));
+
+        if (requestURIParts.size() != 8) {
+            response.sendError(SC_BAD_REQUEST, format("Unexpected URI for task result request in async mode: %s", requestURI));
+            return;
+        }
+
+        TaskId taskId = TaskId.valueOf(requestURIParts.get(4));
+        OutputBufferId bufferId = OutputBufferId.fromString(requestURIParts.get(6));
+        long token = parseLong(requestURIParts.get(7));
+        DataSize maxSize = DataSize.valueOf(request.getHeader(PRESTO_MAX_SIZE));
+
+        AsyncContext asyncContext = request.startAsync(request, response);
+
+        Duration waitTime = randomizeWaitTime(DEFAULT_MAX_WAIT_TIME);
+        asyncContext.setTimeout(waitTime.toMillis() + PAGE_RESPONSE_WAIT_TIME.toMillis());
+
+        // set up async listener
+        asyncContext.addListener(new AsyncListener()
+        {
+            public void onComplete(AsyncEvent event)
+            {
+            }
+
+            public void onError(AsyncEvent event)
+                    throws IOException
+            {
+                String errorMessage = format("Server error to process task result request %s : %s", requestURI, event.getThrowable().getMessage());
+                log.error(event.getThrowable(), errorMessage);
+                response.sendError(SC_INTERNAL_SERVER_ERROR, errorMessage);
+            }
+
+            public void onStartAsync(AsyncEvent event)
+            {
+            }
+
+            public void onTimeout(AsyncEvent event)
+                    throws IOException
+            {
+                String errorMessage = format("Server timeout to process task result request: %s", requestURI);
+                log.error(event.getThrowable(), errorMessage);
+                response.sendError(SC_INTERNAL_SERVER_ERROR, errorMessage);
+            }
+        });
+
+        ListenableFuture<BufferResult> bufferResultFuture = taskManager.getTaskResults(taskId, bufferId, token, maxSize);
+        bufferResultFuture = addTimeout(
+                bufferResultFuture,
+                () -> BufferResult.emptyResults(taskManager.getTaskInstanceId(taskId), token, false),
+                waitTime,
+                timeoutExecutor);
+
+        ServletOutputStream out = response.getOutputStream();
+        addCallback(bufferResultFuture, new FutureCallback<BufferResult>()
```

Comment:
nit: Does it make sense to make the `FutureCallback` to be a static private  inner class? (instead of an anonymous class)